CUDADataFormats/Common/interface/CUDAProductBase.h

```diff
@@ -0,0 +1,90 @@
+#ifndef CUDADataFormats_Common_CUDAProductBase_h
+#define CUDADataFormats_Common_CUDAProductBase_h
+
+#include <atomic>
+#include <memory>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedEventPtr.h"
+
+namespace impl {
+  class CUDAScopedContextBase;
+}
+
+/**
+ * Base class for all instantiations of CUDA<T> to hold the
+ * non-T-dependent members.
+ */
+class CUDAProductBase {
+public:
+  CUDAProductBase() = default;  // Needed only for ROOT dictionary generation
+  ~CUDAProductBase();
```

Comment:
Shouldn't this be `virtual` since you have code inheriting from it? Or are you saying the type is never deleted via its base class pointer?

Reply:
It is never deleted via base class pointer. The only point for the base class is to have separate the non-templated code from `CUDAProduct<T>` in the hope of smaller binary.